crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -445,34 +446,31 @@ impl InstancePool {
         instance_index: usize,
         memories: &mut PrimaryMap<DefinedMemoryIndex, Memory>,
     ) {
-        // Decommit any linear memories that were used
+        // Decommit any linear memories that were used.
         for ((def_mem_idx, memory), base) in
             memories.iter_mut().zip(self.memories.get(instance_index))
         {
             let memory = mem::take(memory);
-            assert!(memory.is_static());
-
-            match memory {
-                Memory::Static {
-                    memory_image: Some(mut image),
-                    ..
-                } => {
-                    // If there was any error clearing the image, just
-                    // drop it here, and let the drop handler for the
-                    // slot unmap in a way that retains the
-                    // address space reservation.
+            let memory_as_any = memory.0.into_any();
+            if let Ok(m) = memory_as_any.downcast::<ExternalMemory>() {
```

Comment:
Putting the `downcast` here and dealing with `ExternalMemory` I think would be better inside of `memory.rs` because otherwise it's created via `Memory::new_static` but then open-coded here that we're reaching into the internals.

Perhaps something like `Memory::unwrap_static_image` which returns `Option<MemoryImage>` which is used here to call `clear_and_remain_ready` and such?

Reply:
I'd personally prefer to make "external" memory an internal implementation detail of the `memory.rs` module if we can. Right now it has a `Memory::new_static` constructor so in theory we would have something like `Memory::unwrap_static_image` where we keep the "static" name and the `unwrap` bit implies the assertion and the method takes out the input argument to `Memory::new_static`

Reply:
Ok, I changed all of the `pub` functions to use the same naming: `is_static` and `unwrap_static_image` now match `new_static`. I left the struct as `ExternalMemory` (instead of `StaticMemory`, though I can change it to that if so desired) because I started to feel that this static/dynamic nomenclature here isn't actually the most accurate. Sure, the old `Memory::Static` had a static starting address, but it turns out `Memory::Dynamic` could as well (e.g., `MemoryStyle::Static`)! Static/dynamic are rather overloaded terms anyways, so I started thinking of this structure in terms of external/internal. External memory is managed "outside the engine" by, e.g., the pooling allocator, and internal memory is managed directly by the engine; this mental switch helped me understand more clearly why we need more than one option here and perhaps it would be the same for others. What do you think? If you agree, I can take up this renaming in a separate PR.